src/native/libs/System.Globalization.Native/pal_collation.m

```diff
@@ -279,4 +280,82 @@ int32_t GlobalizationNative_EndsWithNative(const uint16_t* localeName, int32_t l
     return result == NSOrderedSame ? 1 : 0;
 }
 
+/**
+ * Append a code point to a string, overwriting 1 or 2 code units.
+ * The offset points to the current end of the string contents
+ * and is advanced (post-increment).
+ * "Safe" macro, checks for a valid code point.
+ * If a surrogate pair is written, checks for sufficient space in the string.
+ * If the code point is not valid or a trail surrogate does not fit,
+ * then isError is set to true.
+ *
+ * @param buffer const uint16_t * string buffer
+ * @param offset string offset, must be offset<capacity
+ * @param capacity size of the string buffer
+ * @param codePoint code point to append
+ * @param isError output bool set to true if an error occurs, otherwise not modified
+ */
+#define Append(buffer, offset, capacity, codePoint, isError) { \
+    if ((uint32_t)(codePoint) <= 0xffff) { \
+        (buffer)[(offset)++] = (uint16_t)(codePoint); \
+    } else if ((uint32_t)(codePoint) <= 0x10ffff && (offset) + 1 < (capacity)) { \
+        (buffer)[(offset)++] = (uint16_t)(((codePoint) >> 10) + 0xd7c0); \
+        (buffer)[(offset)++] = (uint16_t)(((codePoint)&0x3ff) | 0xdc00); \
+    } else /* codePoint > 0x10ffff or not enough space */ { \
+        (isError) = true; \
+    } \
```

Comment:
Maybe we can bubble the specific error back up so we can be more specific about the error thrown in `TextInfo.OSX.cs`, instead of using a bool, maybe an int or something.
```suggestion
    } else if ((uint32_t)(codePoint) > 0x10ffff /* invalid code point */  { \
        (isError) = 1; \
    } else if (offset) + 1 >= (capacity)) /* insufficiently sized destination buffer */ { \
        (isError) = 2; \
    } else { \
        (buffer)[(offset)++] = (uint16_t)(((codePoint) >> 10) + 0xd7c0); \
        (buffer)[(offset)++] = (uint16_t)(((codePoint)&0x3ff) | 0xdc00); \
    }
```

Reply:
Updated. To keep current logic of ICU ChangeCase function will not return error in case of insufficiently sized destination buffer but will return same string without casing.

Reply:
Doesn't the ChangeCase function assert that https://github.com/dotnet/runtime/blob/5ca682626651957080b6b4c42cb2ee1c5e743eb3/src/native/libs/System.Globalization.Native/pal_casing.c#L50
in which case, if `isError` is ever true (which it would be for insufficient capacity as well), it would abort?

Reply:
Yes that will abort. I updated the function to return exception in that case.